handle collections as return types or error

if one use List, Set, Map etc. as return types the ask to genAI is: `You must answer strictly in the following JSON format: {\n}` - it should have the the json structure.

This fails:

```
    @RegisterAiService
    public interface jaidataAI {

        @SystemMessage(
            """
            IDENTITY
            
            You are a superintelligent AI that finds all mentions of data structures within an input and...
        """
        )
        @UserMessage("""
                    {data}
                """)
        List<FileInfo> generate(String data);
    }

    record FileInfo(String filename, String content) {};
```

but if change it to the following it works:

```
     FileInfos generate(String data);
    }
    record FileInfos(List<FileInfo>) {};
    record FileInfo(String filename, String content) {};
```

I would think it should be possible to handle - and if not, throw build error it needs to be a non-collection java type?

FWIW, using `Map` throws an error in 0.20.0 due to LangChain4j changes:

```
Caused by: dev.langchain4j.exception.IllegalConfigurationException: Illegal method return type: java.util.Map<java.lang.String, java.lang.Integer>
    at dev.langchain4j.exception.IllegalConfigurationException.illegalConfiguration(IllegalConfigurationException.java:12)
    at dev.langchain4j.service.output.ServiceOutputParser.validateJsonStructure(ServiceOutputParser.java:135)
    at dev.langchain4j.service.output.ServiceOutputParser.outputFormatInstructions(ServiceOutputParser.java:125)
    at io.quarkiverse.langchain4j.runtime.QuarkusServiceOutputParser.outputFormatInstructions(QuarkusServiceOutputParser.java:19)
    at io.quarkiverse.langchain4j.deployment.AiServicesProcessor.gatherMethodMetadata(AiServicesProcessor.java:1098)
    at io.quarkiverse.langchain4j.deployment.AiServicesProcessor.handleAiServices(AiServicesProcessor.java:932)
    at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
    at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
    at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
    at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
    at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2516)
    at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2495)
    at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1521)
    at java.base/java.lang.Thread.run(Thread.java:1583)
    at org.jboss.threads.JBossThread.run(JBossThread.java:483)
```

--- a/core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java
+++ b/core/deployment/src/main/java/io/quarkiverse/langchain4j/deployment/AiServicesProcessor.java
@@ -73,6 +73,7 @@ import org.objectweb.asm.tree.analysis.AnalyzerException;
 
 import com.fasterxml.jackson.databind.PropertyNamingStrategies;
 
+import dev.langchain4j.exception.UnsupportedFeatureException;
 import dev.langchain4j.guardrail.OutputGuardrail;
 import dev.langchain4j.invocation.InvocationParameters;
 import dev.langchain4j.memory.ChatMemory;
@@ -81,7 +82,6 @@ import dev.langchain4j.model.chat.request.json.JsonSchema;
 import dev.langchain4j.service.IllegalConfigurationException;
 import dev.langchain4j.service.Moderate;
 import dev.langchain4j.service.memory.ChatMemoryAccess;
-import dev.langchain4j.service.output.JsonSchemas;
 import dev.langchain4j.service.output.ServiceOutputParser;
 import dev.langchain4j.service.tool.ToolArgumentsErrorHandler;
 import dev.langchain4j.service.tool.ToolErrorContext;
@@ -2105,11 +2105,24 @@ public class AiServicesProcessor {
         List<MethodParameterInfo> params = method.parameters();
 
         // TODO give user ability to provide custom OutputParser
+        // Prefer ServiceOutputParser.jsonSchema over JsonSchemas.jsonSchemaFrom so collection
+        // return types (List/Set of POJO, String, enum, …) get a proper JSON schema (#943).
+        Optional<JsonSchema> structuredOutputSchema = jsonSchemaFrom(returnType);
         String outputFormatInstructions = "";
-        if (!skipOutputFormatInstructionsPredicate.test(method)) {
-            Optional<JsonSchema> structuredOutputSchema = Optional.empty();
-            if (!returnType.equals(Multi.class)) {
+        if (!skipOutputFormatInstructionsPredicate.test(method) && !returnType.equals(Multi.class)) {
+            try {
                 outputFormatInstructions = SERVICE_OUTPUT_PARSER.outputFormatInstructions(returnType);
+            } catch (RuntimeException e) {
+                // e.g. List/Set of POJO: formatInstructions() is unsupported upstream; JSON schema is.
+                if (structuredOutputSchema.isEmpty()) {
+                    throw illegalConfigurationForMethod(
+                            "AI service method return type '" + returnType
+                                    + "' cannot produce output format instructions or a JSON schema. "
+                                    + "Wrap the collection in a record/POJO (e.g. record Items(List<Item> values) {}) "
+                                    + "or use a supported return type",
+                            method);
+                }
+                outputFormatInstructions = "";
             }
         }
 
@@ -2123,7 +2136,7 @@ public class AiServicesProcessor {
 
         AiServiceMethodCreateInfo.ResponseSchemaInfo responseSchemaInfo = ResponseSchemaInfo.of(generateResponseSchema,
                 systemMessageInfo,
-                userMessageInfo.template(), outputFormatInstructions, jsonSchemaFrom(returnType));
+                userMessageInfo.template(), outputFormatInstructions, structuredOutputSchema);
 
         if (!generateResponseSchema && responseSchemaInfo.isInSystemMessage())
             throw new RuntimeException(
@@ -2231,7 +2244,14 @@ public class AiServicesProcessor {
         if (isMulti(returnType)) {
             return Optional.empty();
         }
-        return JsonSchemas.jsonSchemaFrom(returnType);
+        // ServiceOutputParser supports collections; JsonSchemas.jsonSchemaFrom only handles POJOs.
+        // Some AI-service return types (e.g. agentic AgenticScope) are not schema-mappable — treat
+        // that as "no schema" rather than failing the whole build (#943 must not break agentic/mcp).
+        try {
+            return SERVICE_OUTPUT_PARSER.jsonSchema(returnType);
+        } catch (UnsupportedFeatureException | IllegalArgumentException | IllegalStateException e) {
+            return Optional.empty();
+        }
     }
 
     private boolean detectIfToolExecutionRequiresAWorkerThread(MethodInfo method, List<ToolMethodBuildItem> tools,

--- a/model-providers/openai/openai-vanilla/deployment/src/test/java/org/acme/examples/aiservices/CollectionStructuredOutputResponseTest.java
+++ b/model-providers/openai/openai-vanilla/deployment/src/test/java/org/acme/examples/aiservices/CollectionStructuredOutputResponseTest.java
@@ -1,0 +1,85 @@
+package org.acme.examples.aiservices;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.InstanceOfAssertFactories.map;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import dev.langchain4j.service.UserMessage;
+import io.quarkiverse.langchain4j.RegisterAiService;
+import io.quarkiverse.langchain4j.openai.testing.internal.OpenAiBaseTest;
+import io.quarkiverse.langchain4j.testing.internal.WiremockAware;
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * List/Set AI service return types must produce a real JSON schema (not empty {}), see #943.
+ */
+public class CollectionStructuredOutputResponseTest extends OpenAiBaseTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
+            .setArchiveProducer(
+                    () -> ShrinkWrap.create(JavaArchive.class))
+            .overrideRuntimeConfigKey("quarkus.langchain4j.openai.api-key", "whatever")
+            .overrideRuntimeConfigKey("quarkus.langchain4j.openai.chat-model.response-format", "json_schema")
+            .overrideRuntimeConfigKey("quarkus.langchain4j.openai.chat-model.strict-json-schema", "true")
+            .overrideRuntimeConfigKey("quarkus.langchain4j.openai.base-url",
+                    WiremockAware.wiremockUrlForConfig("/v1"));
+
+    record FileInfo(String filename, String content) {
+    }
+
+    @RegisterAiService
+    @ApplicationScoped
+    interface FileInfoExtractor {
+
+        @UserMessage("Find data structure mentions in {{it}}")
+        List<FileInfo> generate(String data);
+    }
+
+    @Inject
+    FileInfoExtractor fileInfoExtractor;
+
+    @Test
+    public void testListOfPojo() throws IOException {
+        setChatCompletionMessageContent(
+                "{\\n\\\"values\\\": [{\\n\\\"filename\\\": \\\"a.txt\\\",\\n\\\"content\\\": \\\"hello\\\"\\n}]\\n}");
+
+        List<FileInfo> result = fileInfoExtractor.generate("some input");
+
+        assertThat(result).hasSize(1);
+        assertThat(result.get(0).filename()).isEqualTo("a.txt");
+        assertThat(result.get(0).content()).isEqualTo("hello");
+
+        Map<String, Object> requestAsMap = getRequestAsMap();
+        assertThat(requestAsMap).hasEntrySatisfying("response_format", (v) -> {
+            assertThat(v).asInstanceOf(map(String.class, Object.class)).satisfies(responseFormatMap -> {
+                assertThat(responseFormatMap).containsEntry("type", "json_schema");
+                assertThat(responseFormatMap).extracting("json_schema").satisfies(js -> {
+                    assertThat(js).asInstanceOf(map(String.class, Object.class)).satisfies(jsonSchemaMap -> {
+                        assertThat(jsonSchemaMap).containsEntry("name", "List_of_FileInfo");
+                        assertThat(jsonSchemaMap).extracting("schema").asInstanceOf(map(String.class, Object.class))
+                                .satisfies(schema -> {
+                                    assertThat(schema).containsKey("properties");
+                                    assertThat(schema.get("properties")).asInstanceOf(map(String.class, Object.class))
+                                            .containsKey("values");
+                                    assertThat(schema.get("properties")).asInstanceOf(map(String.class, Object.class))
+                                            .extracting("values").asInstanceOf(map(String.class, Object.class))
+                                            .containsEntry("type", "array");
+                                });
+                    });
+                });
+            });
+        });
+    }
+}

